File: src/main/java/sprykerFileCreator/model/resolver/ClassResolver.java
package pav.sprykerFileCreator.model.resolver;

import pav.sprykerFileCreator.model.definition.ClassDefinitionInterface;
import pav.sprykerFileCreator.model.definition.DefinitionProviderInterface;
import pav.sprykerFileCreator.model.finder.ClassFinderInterface;
import pav.sprykerFileCreator.model.generator.SprykerConstants;
import pav.sprykerFileCreator.model.renderer.dto.PhpClassInterface;

public class ClassResolver implements ClassResolverInterface {

    private String projectName;
    private String[] projectHierarchy;
    private ClassFinderInterface classFinder;
    private DefinitionProviderInterface definitionProvider;

    public ClassResolver(
            String projectName,
            String[] projectHierarchy,
            ClassFinderInterface classFinder,
            DefinitionProviderInterface definitionProvider
    ) {
        this.projectName = projectName;
        this.projectHierarchy = projectHierarchy;
        this.classFinder = classFinder;
        this.definitionProvider = definitionProvider;
    }

    @Override
    public PhpClassInterface resolveBundleClass(String classType, String bundleName) throws Exception {
        ClassDefinitionInterface classDefinition = this.definitionProvider.getDefinitionByType(classType);
        String classPattern = classDefinition.getNamespacePattern() + "\\" + classDefinition.getNamePattern();
        String fullQualifiedName = classPattern
                .replace(SprykerConstants.PROJECT_NAME_PLACEHOLDER, this.projectName)
                .replace(SprykerConstants.BUNDLE_NAME_PLACEHOLDER, bundleName);

        return this.resolveClass(fullQualifiedName, false, true);
    }

    @Override
    public PhpClassInterface resolveClassBelow(String fullQualifiedName) {
        return this.resolveClass(fullQualifiedName, true, false);
    }

    @Override
    public PhpClassInterface resolveClass(String fullQualifiedName, Boolean findClassBelow, Boolean findInterface) {

        for (String projectBelow : this.projectHierarchy) {

            if (this.projectName.equals(projectBelow) && findClassBelow) {
                continue;
            }

            String fullQualifiedCoreName = fullQualifiedName.replaceFirst(this.projectName, projectBelow);

            if (findInterface) {
                PhpClassInterface coreInterface = this.classFinder.findClass(fullQualifiedCoreName + "Interface");
                if (coreInterface != null) {
                    return coreInterface;
                }
            }

            PhpClassInterface coreClass = this.classFinder.findClass(fullQualifiedCoreName);
            if (coreClass != null) {
                return coreClass;
            }
        }

        return null;
    }

}

File: src/main/kotlin/com/github/tobi812/sprykerplugin/models/resolver/ClassResolver.kt
package com.github.tobi812.sprykerplugin.models.resolver

import com.github.tobi812.sprykerplugin.constants.SprykerConstants
import com.github.tobi812.sprykerplugin.models.definitions.ClassDefinitionInterface
import com.github.tobi812.sprykerplugin.models.definitions.DefinitionProviderInterface
import com.github.tobi812.sprykerplugin.models.finder.ClassFinderInterface
import com.github.tobi812.sprykerplugin.models.renderer.dto.PhpClassInterface
import java.lang.Exception

class ClassResolver(
    private val projectName: String,
    private val projectHierarchy: Array<String>,
    private val classFinder: ClassFinderInterface,
    private val definitionProvider: DefinitionProviderInterface
) : ClassResolverInterface {

    @Throws(Exception::class)
    override fun resolveBundleClass(classType: String, bundleName: String): PhpClassInterface? {
        val classDefinition: ClassDefinitionInterface = definitionProvider.getDefinitionByType(classType)
        val classPattern: String =
            classDefinition.namespacePattern + "\\" + classDefinition.namePattern
        val fullQualifiedName: String = classPattern
            .replace(SprykerConstants.PROJECT_NAME_PLACEHOLDER, projectName)
            .replace(SprykerConstants.MODULE_NAME_PLACEHOLDER, bundleName)

        return resolveClass(fullQualifiedName, findClassBelow = false, findInterface = true)
    }

    override fun resolveClassBelow(fullQualifiedName: String): PhpClassInterface? {
        return this.resolveClass(fullQualifiedName, findClassBelow = true, findInterface = false)
    }

    override fun resolveClass(
        fullQualifiedName: String,
        findClassBelow: Boolean,
        findInterface: Boolean
    ): PhpClassInterface? {
        for (projectBelow in this.projectHierarchy) {
            if (this.projectName == projectBelow && findClassBelow) {
                continue
            }

            val fullQualifiedCoreName = fullQualifiedName.replaceFirst(projectName.toRegex(), projectBelow)
            if (findInterface) {
                val coreInterface: PhpClassInterface? = this.classFinder.findClass(fullQualifiedCoreName + "Interface")
                if (coreInterface != null) {
                    return coreInterface
                }
            }

            val coreClass: PhpClassInterface? = this.classFinder.findClass(fullQualifiedCoreName)
            if (coreClass != null) {
                return coreClass
            }
        }

        return null
    }
}